libraries/GCS_MAVLink/GCS_Signing.cpp

```diff
@@ -0,0 +1,210 @@
+// -*- tab-width: 4; Mode: C++; c-basic-offset: 4; indent-tabs-mode: nil -*-
+
+/*
+  Code for handling MAVLink2 signing
+
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "GCS.h"
+
+extern const AP_HAL::HAL& hal;
+
+#if MAVLINK_PROTOCOL_VERSION >= 2
+
+// storage object
+StorageAccess GCS_MAVLINK::_signing_storage(StorageManager::StorageKeys);
+
+// magic for versioning of the structure
+#define SIGNING_KEY_MAGIC 0x3852fcd1
+
+// structure stored in FRAM
+struct SigningKey {
+    uint32_t magic;
+    uint64_t timestamp;
+    uint8_t secret_key[32];
+};
+
+// shared signing_streams structure
+mavlink_signing_streams_t GCS_MAVLINK::signing_streams;
+
+bool GCS_MAVLINK::signing_key_save(const struct SigningKey &key)
+{
+    if (_signing_storage.size() < sizeof(key)) {
+        return false;
+    }
+    return _signing_storage.write_block(0, &key, sizeof(key));
+}
+
+bool GCS_MAVLINK::signing_key_load(struct SigningKey &key)
+{
+    if (_signing_storage.size() < sizeof(key)) {
+        return false;
+    }
+    return _signing_storage.read_block(&key, 0, sizeof(key));
+}
+
+/*
+  handle a setup_signing message
+ */
+void GCS_MAVLINK::handle_setup_signing(const mavlink_message_t *msg)
```

Comment:
It is a good point. For the moment I think this should be a documentation issue, and we should allow updating of the key over any channel, just for user convenience, but it is worth discussing.
We could for example give a GCS popup warning if the user tries to send this over a signed channel